### PR TITLE
Update DAO to use @gma.aspect.ingestion annotation

### PR DESCRIPTION
**Pull request by zhixuanjia**
With this PR, DAO will read from @gma.aspect.ingestion annotation and decide whether to update the database.

Also moved the UrnPathExtractor out of the legacy `scsi` package. This is a breaking change that will some code change in downstream services.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 1]
```diff
@@ -1520,5 +1517,66 @@ protected boolean aspectVersionSkipWrite(@Nullable RecordTemplate newValue, @Nul
     return aspectVersionComparator(newValue, oldValue) == -1;
   }
 
+  /**
+   * The logic determines if we will update the aspect.
+   */
+  private <ASPECT extends RecordTemplate> boolean  shouldUpdateAspect(IngestionMode ingestionMode, URN urn, ASPECT oldValue,
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:1523]: nit: extra space after boolean

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 2]
```diff
@@ -1520,5 +1517,66 @@ protected boolean aspectVersionSkipWrite(@Nullable RecordTemplate newValue, @Nul
...
+      // If there are filters in annotation, every filter conditions has to be met.
+      boolean passAllFilters = true;
+      for (UrnFilter filter : filters) {
+        if (!urnPaths.containsKey(filter.getPath())) {
+          passAllFilters = false;
+          break;
+        }
+
+        if (!urnPaths.get(filter.getPath()).toString().equals(filter.getValue())) {
+          System.out.println("v: " + filter.getValue());
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:1568]: forgot to remove?

File: gradle-plugins/metadata-annotations-schema/src/main/pegasus/com/linkedin/metadata/annotations/AspectIngestionAnnotation.pdl [hunk 3]
```diff
@@ -23,7 +23,7 @@ record AspectIngestionAnnotation {
   urn: optional string
 
   /**
-   * Filter on the URN so that this ingestion mode is only applicable to a subset of entities.
+   * Filter on the URN so that this ingestion mode is only applicable to a subset of entities. Filters are chained by AND relation.
    */
   filter: optional array[record UrnFilter {
```
Comment (jsdonn) [inline gradle-plugins/metadata-annotations-schema/src/main/pegasus/com/linkedin/metadata/annotations/AspectIngestionAnnotation.pdl:28]: should be OR right?

/platform/platformName = hdfs OR /platform/platformName = oracle

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 4]
```diff
@@ -1520,5 +1517,66 @@ protected boolean aspectVersionSkipWrite(@Nullable RecordTemplate newValue, @Nul
...
+        log.info((String.format("@gma.aspect.ingestion is FORCE_UPDATE on aspect %s and no filters set in annotation."
+            + " Force update aspect.", aspectClass.getCanonicalName())));
+        return true;
+      }
+
+      UrnFilterArray filters = annotation.getFilter();
+      Map<String, Object> urnPaths = _urnPathExtractor.extractPaths(urn);
+
+      // If there are filters in annotation, every filter conditions has to be met.
+      boolean passAllFilters = true;
```
Comment (jsdonn) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:1560]: I feel like it should be an OR not AND?

or else how do we specify multiple data platforms?

File: testing/test-models/src/main/pegasus/com/linkedin/testing/AspectFoo.pdl [hunk 5]
```diff
@@ -3,6 +3,11 @@ namespace com.linkedin.testing
 /**
  * For unit tests
  */
+@gma.aspect.ingestion = [
```
Comment (jsdonn) [inline testing/test-models/src/main/pegasus/com/linkedin/testing/AspectFoo.pdl:6]: can you pls writ a test with multiple filters? maybe 1 for multiple filters on different paths, 1 for multiple filters on the same path

**Review by jsdonn — COMMENTED** [review]
some small comments, major comment is I think filters should be treated with OR not AND, lmk your thoughts.

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 1 above)
Comment (zhixuanjia) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: fixed

File: gradle-plugins/metadata-annotations-schema/src/main/pegasus/com/linkedin/metadata/annotations/AspectIngestionAnnotation.pdl (hunk 3 above)
Comment (zhixuanjia) [reply gradle-plugins/metadata-annotations-schema/src/main/pegasus/com/linkedin/metadata/annotations/AspectIngestionAnnotation.pdl:28]: you are right, change to OR

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 4 above)
Comment (zhixuanjia) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java]: changed to OR

File: testing/test-models/src/main/pegasus/com/linkedin/testing/AspectFoo.pdl (hunk 5 above)
Comment (zhixuanjia) [reply testing/test-models/src/main/pegasus/com/linkedin/testing/AspectFoo.pdl:6]: add two more test cases to cover multiple filter case

**Review by jsdonn — APPROVED** [review]
LGTM, thanks for helping with this and adding the additional test cases

**Review by jsdonn — APPROVED** [review]
LGTM thanks!

**Pull request merged**